stl/src/sharedmutex.cpp

```diff
@@ -36,4 +36,9 @@ void __cdecl _Smtx_unlock_exclusive(_Smtx_t* smtx) { // unlock exclusive shared
 void __cdecl _Smtx_unlock_shared(_Smtx_t* smtx) { // unlock non-exclusive shared mutex
     ReleaseSRWLockShared(reinterpret_cast<PSRWLOCK>(smtx));
 }
+
+void __cdecl _Thrd_sleep_for(const long long ns) { // attempt to suspend thread for at least `ns` nanoseconds
+    constexpr long nsec_per_msec = 1'000'000L;
+    Sleep(static_cast<DWORD>((ns + nsec_per_msec - 1) / nsec_per_msec));

```

Comment:
It will require preprocessor pain but you can reduce this problem to an `NtWaitForSingleObject` call which will give you more range and the ability to reasonably handle absolute waits against the system clock. Here's the code I wrote way back when we were going to do an ABI break in 2017... please excuse the Whitesmiths indent style and the `__vcp_` prefix on things back when we thought VCPlatform would be a thing.

The trick is to wait on `GetCurrentThread()`; we know the current thread can never be signaled, because it is running.

```
constexpr unsigned long _Max_sleep_length{ULONG_MAX - 1U};

static void _Bounds_check(const unsigned long long _Relative_filetime) noexcept
    {    // checks that the input FILETIME is in valid bounds
    if (_Relative_filetime > static_cast<unsigned long long>(LLONG_MAX))
        {    // out of range (~30,000 years in the future)
        __vcp_fastfail();
        }
    }

static void _Win32_sleep(unsigned long long _Relative_filetime) noexcept
    {    // low resolution relative sleep in terms of Win32 Sleep
    _Relative_filetime /= 10'000;    // convert from 100ns intervals to 1ms intervals
                                    // 1e-7 -> 1e-3 == 4 zeros
    while (_Relative_filetime > _Max_sleep_length)
        {
        Sleep(_Max_sleep_length);
        _Relative_filetime -= _Max_sleep_length;
        }

    Sleep(static_cast<unsigned long>(_Relative_filetime));
    }

extern "C" void __cdecl __std_thread_sleep_system(const unsigned long long _Relative_filetime) noexcept
    {    // sleep a relative time measured against the system clock
    _Bounds_check(_Relative_filetime);
    if (__vcp_has_NtWaitForSingleObject())
        {
        long long _Now;
        __vcp_get_system_time_precise_as_file_time(reinterpret_cast<FILETIME *>(&_Now));
        const long long _Offset = static_cast<long long>(_Relative_filetime); // safe because we _Bounds_check'ed
        if (_Now > LLONG_MAX - _Offset)
            {    // out of range
            __vcp_fastfail();
            }

        LARGE_INTEGER _Desired;
        _Desired.QuadPart = _Now + _Offset;
        __vcp_NtWaitForSingleObject(
            GetCurrentThread(), // Can't become signaled because we're running
            FALSE,
            &_Desired // positive value == absolute wait
            );
        }
    else
        {    // can't do absolute waits on this platform, fall back to relative
        _Win32_sleep(_Relative_filetime);
        }
    }

extern "C" void __cdecl __std_thread_sleep_steady(const unsigned long long _Relative_filetime) noexcept
    {    // sleep a relative time measured against a steady clock
    _Bounds_check(_Relative_filetime);
    if (__vcp_has_NtWaitForSingleObject())
        {
        LARGE_INTEGER _Desired;
        _Desired.QuadPart = -static_cast<long long>(_Relative_filetime);
        __vcp_NtWaitForSingleObject(
            GetCurrentThread(), // Can't become signaled because we're running
            FALSE,
            &_Desired // negative value == relative wait
            );
        }
    else
        {
        _Win32_sleep(_Relative_filetime);
        }
    }

template<class _Rep,
    class _Period> inline
    long long _Time_to_filetime(const chrono::duration<_Rep, _Period>& _Rel_time)
    {    // convert a relative time to the minimum time greater than or equal (so that we wait for at least _Rel_time)
    return (chrono::ceil<
        chrono::duration<long long, ratio<1, 10'000'000>>>(_Rel_time).count());
    }

template<class _Rep,
    class _Period> inline
    void sleep_for(const chrono::duration<_Rep, _Period>& _Rel_time)
    {    // sleep for duration
    if (_Rel_time > chrono::duration<_Rep, _Period>::zero())
        {
        __std_thread_sleep_steady(_Time_to_filetime(_Rel_time));
        }
    }

template<class _Clock,
    class _Duration> inline
    void sleep_until(const chrono::time_point<_Clock, _Duration>& _Abs_time)
    {    // sleep until time point
    const auto _Abs_target = _Abs_time.time_since_epoch();
    for (;;)
        {
        const auto _Now = _Clock::now().time_since_epoch();
        if (!(_Now < _Abs_target))
            return;
        const auto _Relative_filetime = _Time_to_filetime(_Abs_target - _Now);
        if (is_same_v<_Clock, chrono::system_clock>)
            {
            __std_thread_sleep_system(_Relative_filetime);
            }
        else
            {
            __std_thread_sleep_steady(_Relative_filetime);
            }
        }
    }
```

Reply:
I'm going to resolve this without taking action because that's a ton of code and I'm not sure if we can use `NtWaitForSingleObject` (it's desktop-only, has to be GetProcAddressed, etc).